aiida/backends/tests/tools/graph/test_age.py

```diff
@@ -0,0 +1,738 @@
+
+        basket = Basket(nodes=[data_node.id])
+        queryb = orm.QueryBuilder()
+        queryb.append(orm.Node).append(orm.Node)
+
+        uprule = UpdateRule(queryb, max_iterations=np.inf)
+        obtained = uprule.run(basket.copy())['nodes'].keyset
+        expected = set([data_node.id, work_node.id])
+        self.assertEqual(obtained, expected)
+
+        rerule1 = ReplaceRule(queryb, max_iterations=1)
+        result1 = rerule1.run(basket.copy())['nodes'].keyset
+        self.assertEqual(result1, set([work_node.id]))
+
+        rerule2 = ReplaceRule(queryb, max_iterations=2)
+        result2 = rerule2.run(basket.copy())['nodes'].keyset
+        self.assertEqual(result2, set([data_node.id]))
+
+        rerule3 = ReplaceRule(queryb, max_iterations=3)
+        result3 = rerule3.run(basket.copy())['nodes'].keyset
+        self.assertEqual(result3, set([work_node.id]))
+
+        rerule4 = ReplaceRule(queryb, max_iterations=4)
+        result4 = rerule4.run(basket.copy())['nodes'].keyset
+        self.assertEqual(result4, set([data_node.id]))
+
+    @staticmethod
+    def _create_branchy_graph():
+        """
+        Creates a basic branchy graph which has two concatenated calculations:
+
+        * calc_1 takes data_0 as an input and returns data_1 and data_o.
+        * calc_2 takes data_1 and data_i as inputs and returns data_2.
+
+        This graph looks like this::
+
+                           +--------+                    +--------+
+                           | data_o |                    | data_i |
+                           +--------+                    +--------+
+                               ^                             |
+                               |                             v
+            +--------+     +--------+     +--------+     +--------+     +--------+
+            | data_0 | --> | calc_1 | --> | data_1 | --> | calc_2 | --> | data_2 |
+            +--------+     +--------+     +--------+     +--------+     +--------+
+        """
+        data_0 = orm.Data().store()
+        calc_1 = orm.CalculationNode()
+        calc_1.add_incoming(data_0, link_type=LinkType.INPUT_CALC, link_label='inpcalc_data_0')
+        calc_1.store()
+
+        data_1 = orm.Data()
+        data_o = orm.Data()
+        data_1.add_incoming(calc_1, link_type=LinkType.CREATE, link_label='create_data_1')
+        data_o.add_incoming(calc_1, link_type=LinkType.CREATE, link_label='create_data_o')
+        data_1.store()
+        data_o.store()
+
+        data_i = orm.Data().store()
+        calc_2 = orm.CalculationNode()
+        calc_2.add_incoming(data_1, link_type=LinkType.INPUT_CALC, link_label='inpcalc_data_1')
+        calc_2.add_incoming(data_i, link_type=LinkType.INPUT_CALC, link_label='inpcalc_data_i')
+        calc_2.store()
+
+        data_2 = orm.Data()
+        data_2.add_incoming(calc_2, link_type=LinkType.CREATE, link_label='create_data_2')
+        data_2.store()
+
+        output_dict = {}
```

Comment:
I'm still in favour of making this a one-liner:
```suggestion
        output_dict = {
            'data_i': data_i,
            'data_0': data_0,
            'data_1': data_1,
            'data_2': data_2,
            'data_o': data_o,
            'calc_1': calc_1,
            'calc_2': calc_2
        }
```